File: verifiers/GachaResolve.java
// ============================================================================
// UVS uvGacha - REFERENCE RESOLVER (Java, stdlib only). Mirrors gacha-resolve.js.
//   combinedSeed = SHA-256( serverSeed : clientSeed : drandRandomness )
//   u_i          = SHA-256( combinedSeed : i )  as a 256-bit big-endian int, mod D  (BigInteger)
//   outcome      = the tier whose cumulative integer interval in [0,D) contains u_i
// Optional hard-pity (uvGacha 5) is reconstructed by replay from pull 1, never stored (core 6.1).
//
// Run:  java GachaResolve.java <record.json>     (Java 11+; no javac needed)
//       java GachaResolve.java                   (self-test against test-vectors-gacha.json)
// ============================================================================
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.security.MessageDigest;
import java.util.*;

public class GachaResolve {

    static String sha256(String s) throws Exception {
        byte[] d = MessageDigest.getInstance("SHA-256").digest(s.getBytes(StandardCharsets.UTF_8));
        StringBuilder sb = new StringBuilder();
        for (byte b : d) sb.append(String.format("%02x", b));
        return sb.toString();
    }
    static String combinedSeed(String server, String client, String drand) throws Exception {
        return sha256(server + ":" + client + ":" + (drand == null ? "" : drand));
    }
    // u_i = SHA-256(combined:i) as a 256-bit big-endian integer, mod D
    static long pullValue(String combined, int i, long D) throws Exception {
        return new BigInteger(sha256(combined + ":" + i), 16).mod(BigInteger.valueOf(D)).longValue();
    }

    @SuppressWarnings("unchecked")
    static void validateRules(Map<String,Object> rules, long D) {
        if (D <= 0) throw new IllegalArgumentException("INVALID: rateDenominator must be a positive integer (uvGacha 4)");
        long sum = 0;
        for (Object o : (List<Object>) rules.get("tiers")) {
            Map<String,Object> t = (Map<String,Object>) o;
            if (!(t.get("tier") instanceof String)) throw new IllegalArgumentException("INVALID: tier label must be a string (uvGacha 4)");
            if (!(t.get("rate") instanceof Long)) throw new IllegalArgumentException("INVALID: rate must be a JSON integer (uvGacha 4)");
            long r = (Long) t.get("rate");
            if (r < 0) throw new IllegalArgumentException("INVALID: rate must be non-negative (uvGacha 4)");
            sum += r;
        }
        if (sum != D) throw new IllegalArgumentException("INVALID: rates sum to " + sum + " != D " + D + " (uvGacha 4)");
    }
    @SuppressWarnings("unchecked")
    static String tierOf(List<Object> tiers, long u) {
        long acc = 0;
        for (Object o : tiers) { Map<String,Object> t = (Map<String,Object>) o; acc += (Long) t.get("rate"); if (u < acc) return (String) t.get("tier"); }
        throw new IllegalArgumentException("INVALID: u beyond cumulative range");
    }

    @SuppressWarnings("unchecked")
    static List<Map<String,Object>> resolve(Map<String,Object> rec) throws Exception {
        long D = (Long) rec.get("rateDenominator");
        Map<String,Object> rules = (Map<String,Object>) rec.get("rules");
        validateRules(rules, D);
        Object dr = rec.get("drand");
        String drand = (dr instanceof Map) ? (String) ((Map<String,Object>) dr).get("randomness") : null;
        String combined = combinedSeed((String) rec.get("serverSeed"), (String) rec.get("clientSeed"), drand);
        Map<String,Object> pity = rules.get("pity") instanceof Map ? (Map<String,Object>) rules.get("pity") : null;
        List<Object> tiers = (List<Object>) rules.get("tiers");
        long pullCount = (Long) rec.get("pullCount");
        List<Map<String,Object>> out = new ArrayList<>();
        int miss = 0;
        for (int i = 1; i <= pullCount; i++) {
            long u = pullValue(combined, i, D);
            String tier; boolean forced = false;
            if (pity != null && miss + 1 >= (Long) pity.get("hardAfter")) { tier = (String) pity.get("tier"); forced = true; }
            else tier = tierOf(tiers, u);
            Map<String,Object> r = new LinkedHashMap<>(); r.put("i", (long) i); r.put("tier", tier); r.put("forced", forced);
            out.add(r);
            String pt = pity != null ? (String) pity.get("tier") : null;
            miss = (pity != null && tier.equals(pt)) ? 0 : miss + 1;
        }
        out.add(0, mapOf("combinedSeed", combined));  // index 0 carries the combinedSeed for the caller
        return out;
    }
    static Map<String,Object> mapOf(String k, Object v) { Map<String,Object> m = new LinkedHashMap<>(); m.put(k, v); return m; }

    // ---- minimal JSON parser (objects -> LinkedHashMap, arrays -> ArrayList) ----
    static final class P {
        final String s; int i = 0;
        P(String s) { this.s = s; }
        void ws() { while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++; }
        Object value() {
            ws(); char ch = s.charAt(i);
            if (ch == '{') return obj(); if (ch == '[') return arr(); if (ch == '"') return str();
            if (ch == 't') { i += 4; return Boolean.TRUE; }
            if (ch == 'f') { i += 5; return Boolean.FALSE; }
            if (ch == 'n') { i += 4; return null; }
            int j = i; while (j < s.length() && "+-0123456789.eE".indexOf(s.charAt(j)) >= 0) j++;
            String n = s.substring(i, j); i = j;
            return n.contains(".") || n.contains("e") || n.contains("E") ? (Object) Double.parseDouble(n) : (Object) Long.parseLong(n);
        }
        Map<String,Object> obj() {
            Map<String,Object> m = new LinkedHashMap<>(); i++; ws();
            if (s.charAt(i) == '}') { i++; return m; }
            while (true) { ws(); String k = str(); ws(); i++; m.put(k, value()); ws();
                if (s.charAt(i) == ',') { i++; continue; } i++; return m; }
        }
        List<Object> arr() {
            List<Object> a = new ArrayList<>(); i++; ws();
            if (s.charAt(i) == ']') { i++; return a; }
            while (true) { a.add(value()); ws();
                if (s.charAt(i) == ',') { i++; continue; } i++; return a; }
        }
        String str() {
            StringBuilder b = new StringBuilder(); i++;
            while (s.charAt(i) != '"') {
                char ch = s.charAt(i++);
                if (ch == '\\') { char e = s.charAt(i++);
                    switch (e) { case 'n': b.append('\n'); break; case 't': b.append('\t'); break;
                        case 'r': b.append('\r'); break; case 'b': b.append('\b'); break; case 'f': b.append('\f'); break;
                        case 'u': b.append((char) Integer.parseInt(s.substring(i, i + 4), 16)); i += 4; break;
                        default: b.append(e); } }
                else b.append(ch);
            }
            i++; return b.toString();
        }
    }
    @SuppressWarnings("unchecked")
    static Map<String,Object> parse(String json) { return (Map<String,Object>) new P(json).value(); }

    @SuppressWarnings("unchecked")
    public static void main(String[] args) throws Exception {
        if (args.length > 0) {
            Map<String,Object> rec = parse(new String(Files.readAllBytes(Paths.get(args[0])), StandardCharsets.UTF_8));
            List<Map<String,Object>> r = resolve(rec);
            System.out.println("combinedSeed = " + r.get(0).get("combinedSeed"));
            for (int k = 1; k < r.size(); k++) System.out.println("  pull " + r.get(k).get("i") + ": " + r.get(k).get("tier") + (((Boolean) r.get(k).get("forced")) ? "  (pity)" : ""));
            return;
        }
        // self-test against the committed vectors
        String json = null;
        for (String p : new String[]{"test-vectors-gacha.json", "verifiers/test-vectors-gacha.json"})
            if (Files.exists(Paths.get(p))) { json = new String(Files.readAllBytes(Paths.get(p)), StandardCharsets.UTF_8); break; }
        if (json == null) { System.out.println("test-vectors-gacha.json not found (run from repo root or verifiers/)"); System.exit(2); }
        Map<String,Object> tv = parse(json);

        Map<String,Object> st = (Map<String,Object>) tv.get("stateless");
        List<Map<String,Object>> A = resolve((Map<String,Object>) st.get("record"));
        boolean aOk = A.get(0).get("combinedSeed").equals(st.get("combinedSeed"));
        List<Object> expTiers = (List<Object>) st.get("tiers");
        for (int k = 0; k < expTiers.size(); k++) if (!A.get(k + 1).get("tier").equals(expTiers.get(k))) aOk = false;

        Map<String,Object> pt = (Map<String,Object>) tv.get("pity");
        List<Map<String,Object>> B = resolve((Map<String,Object>) pt.get("record"));
        boolean bOk = B.get(0).get("combinedSeed").equals(pt.get("combinedSeed"));
        List<Object> expPulls = (List<Object>) pt.get("pulls");
        for (int k = 0; k < expPulls.size(); k++) {
            Map<String,Object> e = (Map<String,Object>) expPulls.get(k), g = B.get(k + 1);
            if (!g.get("tier").equals(e.get("tier")) || !g.get("forced").equals(e.get("forced"))) bOk = false;
        }

        Map<String,Object> neg = (Map<String,Object>) ((Map<String,Object>) tv.get("negative")).get("rates-not-sum-D");
        boolean negOk = false;
        try { resolve((Map<String,Object>) neg.get("record")); } catch (IllegalArgumentException e) { negOk = true; }

        System.out.println("stateless " + aOk + " | pity " + bOk + " | rates-sum-reject " + negOk);
        System.out.println("gacha self-test " + (aOk && bOk && negOk ? "PASS" : "FAIL"));
    }
}

File: verifiers/gacha_resolve.cpp
// ============================================================================
// UVS uvGacha - REFERENCE RESOLVER (C++17, no dependencies)
//
// A gacha session is a deterministic resolver replayed over committed entropy:
//   combinedSeed = SHA-256( serverSeed : clientSeed : drandRandomness )
//   u_i          = SHA-256( combinedSeed : i )  as a 256-bit big-endian int, mod D
//   outcome      = the tier whose cumulative integer interval in [0,D) contains u_i
// The 256-bit value mod D is computed by base-16 nibble iteration -> no big-int needed,
// byte-identical to the JS (BigInt) / Python / Java (BigInteger) resolvers.
// Optional hard-pity (uvGacha 5) is reconstructed by replay from pull 1, never stored.
//
// Build (MSVC):  cl /EHsc /O2 gacha_resolve.cpp     then  gacha_resolve.exe
// Build (g++):   g++ -std=c++17 -O2 -o gacha gacha_resolve.cpp
// (C++ has no stdlib SHA-256, so a compact FIPS-180-4 implementation is bundled.)
// ============================================================================
#define _CRT_SECURE_NO_WARNINGS
#include <cstdint>
#include <cstdio>
#include <string>
#include <vector>
using namespace std;

static const uint32_t K[64] = {
  0x428a2f98,0x71374491,0xb5c0fbcf,0xe9b5dba5,0x3956c25b,0x59f111f1,0x923f82a4,0xab1c5ed5,
  0xd807aa98,0x12835b01,0x243185be,0x550c7dc3,0x72be5d74,0x80deb1fe,0x9bdc06a7,0xc19bf174,
  0xe49b69c1,0xefbe4786,0x0fc19dc6,0x240ca1cc,0x2de92c6f,0x4a7484aa,0x5cb0a9dc,0x76f988da,
  0x983e5152,0xa831c66d,0xb00327c8,0xbf597fc7,0xc6e00bf3,0xd5a79147,0x06ca6351,0x14292967,
  0x27b70a85,0x2e1b2138,0x4d2c6dfc,0x53380d13,0x650a7354,0x766a0abb,0x81c2c92e,0x92722c85,
  0xa2bfe8a1,0xa81a664b,0xc24b8b70,0xc76c51a3,0xd192e819,0xd6990624,0xf40e3585,0x106aa070,
  0x19a4c116,0x1e376c08,0x2748774c,0x34b0bcb5,0x391c0cb3,0x4ed8aa4a,0x5b9cca4f,0x682e6ff3,
  0x748f82ee,0x78a5636f,0x84c87814,0x8cc70208,0x90befffa,0xa4506ceb,0xbef9a3f7,0xc67178f2
};
static inline uint32_t rotr(uint32_t x, int n) { return (x >> n) | (x << (32 - n)); }

string sha256_hex(const string& msg) {
  uint32_t h[8] = {0x6a09e667,0xbb67ae85,0x3c6ef372,0xa54ff53a,0x510e527f,0x9b05688c,0x1f83d9ab,0x5be0cd19};
  vector<uint8_t> d(msg.begin(), msg.end());
  uint64_t bitlen = (uint64_t)d.size() * 8;
  d.push_back(0x80);
  while (d.size() % 64 != 56) d.push_back(0x00);
  for (int i = 7; i >= 0; i--) d.push_back((uint8_t)((bitlen >> (i * 8)) & 0xff));
  for (size_t off = 0; off < d.size(); off += 64) {
    uint32_t w[64];
    for (int i = 0; i < 16; i++)
      w[i] = ((uint32_t)d[off+i*4]<<24)|((uint32_t)d[off+i*4+1]<<16)|((uint32_t)d[off+i*4+2]<<8)|((uint32_t)d[off+i*4+3]);
    for (int i = 16; i < 64; i++) {
      uint32_t s0 = rotr(w[i-15],7)^rotr(w[i-15],18)^(w[i-15]>>3);
      uint32_t s1 = rotr(w[i-2],17)^rotr(w[i-2],19)^(w[i-2]>>10);
      w[i] = w[i-16] + s0 + w[i-7] + s1;
    }
    uint32_t a=h[0],b=h[1],c=h[2],dd=h[3],e=h[4],f=h[5],g=h[6],hh=h[7];
    for (int i = 0; i < 64; i++) {
      uint32_t S1 = rotr(e,6)^rotr(e,11)^rotr(e,25);
      uint32_t ch = (e&f)^((~e)&g);
      uint32_t t1 = hh + S1 + ch + K[i] + w[i];
      uint32_t S0 = rotr(a,2)^rotr(a,13)^rotr(a,22);
      uint32_t maj = (a&b)^(a&c)^(b&c);
      uint32_t t2 = S0 + maj;
      hh=g; g=f; f=e; e=dd+t1; dd=c; c=b; b=a; a=t1+t2;
    }
    h[0]+=a;h[1]+=b;h[2]+=c;h[3]+=dd;h[4]+=e;h[5]+=f;h[6]+=g;h[7]+=hh;
  }
  char out[65];
  for (int i = 0; i < 8; i++) snprintf(out + i*8, 9, "%08x", h[i]);
  out[64] = 0;
  return string(out);
}

string combinedSeed(const string& server, const string& client, const string& drand) {
  return sha256_hex(server + ":" + client + ":" + drand);
}
// u_i = SHA-256(combined:i) as a 256-bit big-endian integer, mod D (base-16 nibble iteration)
int64_t pullValue(const string& combined, int i, int64_t D) {
  string hx = sha256_hex(combined + ":" + to_string(i));
  int64_t acc = 0;
  for (char c : hx) { int v = (c >= '0' && c <= '9') ? c - '0' : c - 'a' + 10; acc = (acc * 16 + v) % D; }
  return acc;
}
struct Tier { string name; int64_t rate; };
// rates MUST be non-negative and sum to exactly D (uvGacha §4)
bool ratesValid(const vector<Tier>& tiers, int64_t D) {
  int64_t sum = 0;
  for (auto& t : tiers) { if (t.rate < 0) return false; sum += t.rate; }
  return sum == D;
}
string tierOf(const vector<Tier>& tiers, int64_t u) {
  int64_t acc = 0;
  for (auto& t : tiers) { acc += t.rate; if (u < acc) return t.name; }
  return "ERR";
}
// resolve a session. hardAfter>0 enables the §5 example hard-pity machine for tier `pityTier`.
vector<pair<string,bool>> resolve(const vector<Tier>& tiers, int64_t D, const string& combined,
                                  int pullCount, const string& pityTier, int hardAfter) {
  vector<pair<string,bool>> out;
  int miss = 0;
  for (int i = 1; i <= pullCount; i++) {
    int64_t u = pullValue(combined, i, D);
    string tier; bool forced = false;
    if (hardAfter > 0 && miss + 1 >= hardAfter) { tier = pityTier; forced = true; }
    else tier = tierOf(tiers, u);
    out.push_back({tier, forced});
    miss = (hardAfter > 0 && tier == pityTier) ? 0 : miss + 1;
  }
  return out;
}

int main() {
  const string ss = "a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8091a2b3c4d5e6f70811223344556";
  const int64_t D = 1000000;
  vector<Tier> tiers = {{"5star",6000},{"4star",91000},{"3star",903000}};

  // -- stateless vector --
  string cA = combinedSeed(ss, "uvs-gacha-demo", "");
  auto A = resolve(tiers, D, cA, 10, "", 0);
  const char* expA[10] = {"4star","3star","3star","3star","3star","3star","3star","3star","3star","3star"};
  bool aOk = (cA == "91d696b5dd3f9634d7015b321685ae61ac3a52eb4cf540cf11d5a1a2cc561043");
  for (int i = 0; i < 10; i++) if (A[i].first != expA[i]) aOk = false;

  // -- pity vector (hard pity: 5star guaranteed on the 10th consecutive miss) --
  string cB = combinedSeed(ss, "uvs-gacha-pity", "");
  auto B = resolve(tiers, D, cB, 12, "5star", 10);
  const char* expB[12] = {"3star","3star","3star","3star","3star","3star","3star","3star","3star","5star","4star","3star"};
  bool bOk = (cB == "fa753fc5786340d20b5857a90736473fb4360f55d2e4a9a3eed6f36909710cdf") && B[9].second; // pull 10 forced
  for (int i = 0; i < 12; i++) if (B[i].first != expB[i]) bOk = false;

  // -- negative: rates 6000+91000+900000 = 997000 != D --
  vector<Tier> bad = {{"5star",6000},{"4star",91000},{"3star",900000}};
  bool negOk = !ratesValid(bad, D) && ratesValid(tiers, D);

  printf("combinedSeed(stateless) = %s\n", cA.c_str());
  printf("stateless %s | pity %s | rates-sum-reject %s\n", aOk?"OK":"FAIL", bOk?"OK":"FAIL", negOk?"OK":"FAIL");
  printf("gacha self-test %s\n", (aOk && bOk && negOk) ? "PASS" : "FAIL");
  return 0;
}
